DOCS.md

```diff
@@ -0,0 +1,246 @@
+- `<return_type> <name>_NEWFRAME(...)` is a version of `<name>_RUN` that halts all workers (between tasks) and runs the `<name>` task on the Lace workers. One use case is stop-the-world garbage collection.
+- `void <name>_TOGETHER(...)` is a version of `<name>_RUN` that lets all workers run the same task concurrently. One use case is initialization of thread local storage of the program on every Lace worker.
+
+Use the `SPAWN` and `SYNC` methods to create tasks that can be stolen and to obtain the result of the last spawned tasks. The number of `SPAWN` and `SYNC` calls should match.
+
+For defining tasks with no return type (`void`), use these macros:
+- `VOID_TASK_1(name, type_1, name_1)`
+- `VOID_TASK_2(name, type_1, name_1, type_2, name_2)`
+- `VOID_TASK_3(name, type_1, name_1, type_2, name_2, type_3, name_3)`
+- etc.
+
+**Important**: the actual task must be a function with a signature like:
+- `<return_type> <name>(LaceWorker*, <type_1>, <type_2>, ...)`
+- `void <name>(LaceWorker*, <type_1>, <type_2>, ...)`
+
+The tasks are given the `LaceWorker*` pointer that corresponds to the worker that the task is currently running in. This pointer and its contents must not be changed and the pointer is necessary for a number of function calls such as `SPAWN`, `SYNC`, etc.
+
+### Example
+
+Typically a recursive function with two 'child tasks' is parallelized as follows:
+
+```c
+TASK_1(int, fibonacci, int, n)   // macro to create Lace functions (can be in header file)
+
+int fibonacci(LaceWorker* lw, int n) {
+    if(n < 2) return n;
+    fibonacci_SPAWN(lw, n-1);    // SPAWN a task (fork)
+    int a = fibonacci(lw, n-2);  // run another task in parallel
+    int b = fibonacci_SYNC(lw);  // SYNC the spawned task (join)
+    return a + b;
+}
+
+int main(int argc, char** argv)
+{
+    int n_workers = 4;           // create 4 workers
+                                 // use 0 to automatically use all available cores
+    int dqsize = 0;              // use default task deque size
+    int stacksize = 0;           // use default program stack size
+
+    lace_start(n_workers, dqsize, stacksize);
+    int result = fibonacci_RUN(42);
+    printf("fibonacci(42) = %d\n", result);
+    lace_stop();
+}
+```
+
+See further the [benchmarks](benchmarks/) directory for more examples.
+
+### Interrupting
+
+Lace offers two methods to interrupt currently running tasks and run something else:
+- the `fib_NEWFRAME(40)` function halts current tasks and offers the `fib(40)` task to the framework.
+- the `fub_TOGETHER(40)` function halts current tasks and lets **all Lace workers** execute a copy of the given `fib(40)` task.
+
+The `TOGETHER` functions are typically used to initialize thread-local variables on each worker, or to perform some other operation on each individual worker thread. The `NEWFRAME` functions are used for example to implement stop-the-world garbage collection, interrupting current tasks in order to run a more important task first.
```

Comment:
Order should probably match the itemization above, i.e. `NEWFRAME` first and then `TOGETHER`.